sklearn/utils/estimator_checks.py

```diff
@@ -1190,6 +1189,32 @@ def check_estimators_pickle(name, estimator_orig):
         unpickled_result = getattr(unpickled_estimator, method)(X)
         assert_allclose_dense_sparse(result[method], unpickled_result)
 
+    # check if input with nans pickles properly
+    if name in ALLOW_NAN:
+        # set random 10 elements to be np.nan
+        rng = np.random.RandomState(42)
+        mask = rng.choice(X.size, 10, replace=False)
+        X.reshape(-1)[mask] = np.nan
+        estimator.fit(X, y)
+
+        nan_result = dict()
+        for method in check_methods:
+            if hasattr(estimator, method):
+                nan_result[method] = getattr(estimator, method)(X)
+
+        # pickle and unpickle!
+        pickled_estimator = pickle.dumps(estimator)
+        unpickled_estimator = pickle.loads(pickled_estimator)
+
+        for method in nan_result:
+            unpickled_nan_result = getattr(unpickled_estimator, method)(X)
+            if name == 'ChainedImputer':
+                tol = {'rtol': 1e-5, 'atol': 0.1}
+            else:
+                tol = {}
+            assert_allclose_dense_sparse(nan_result[method],
```

Comment:
I'd think failure is most likely here, but it may also occur above when calling the method.

But basically, by improving the error message, I just mean that it would be good to have an output that says """Cannot reuse pickled estimator when data contains NaN. Perhaps you should be using isnan() instead of testing for identity."""

Reply:
This is where the error occurs without the patch, as the results between the original and depickled version differ. 

I'm not sure if every time this `assert_allclose_dense_sparse` call fails would be due to this identify problem though. Wouldn't we want to keep the error message the same here? i.e. for ChainedImputer:
```python
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=1e-09
E           
E           x and y nan location mismatch:
E            x: array([[0.769674, 1.328208, 1.268197],
E                  [0.378328, 0.375537, 0.216566],
E                  [0.296359, 0.269703, 0.400726],...
E            y: array([[     nan, 1.328208, 1.268197],
E                  [0.378328, 0.375537, 0.216566],
E                  [0.296359, 0.269703, 0.400726],...
```